Upgrade `openapi-typescript` to 7.0.2 + swap to `openapi-fetch`

davelopez opened the pull request:
Upgrade to the latest version of `openapi-typescript`. This introduces some [breaking changes](https://github.com/openapi-ts/openapi-typescript/releases/tag/openapi-typescript%407.0.0) in particular:

> Most optional objects are now always present in types, just typed as :never. This includes keys of the Components Object as well as HTTP methods.

Maybe this is a good opportunity to continue with @dannon's work started in https://github.com/dannon/galaxy/commits/openapi-typescript-fetch-upgrade/ and swap to [openapi-fetch](https://github.com/openapi-ts/openapi-typescript/tree/main/packages/openapi-fetch)

## Update

See discussion details at https://github.com/galaxyproject/galaxy/discussions/18534

TLDR; There are some more incompatibilities between the latest `openapi-typescript` generated schemas and the current `openapi-typescript-fetcher` so, if we want to upgrade we should replace the current fetcher with the most up-to-date `openapi-fetch` variant.

## TODO
- [x] Replace `fetcher` with `client`
- [x] Improve a bunch of things in the API schema
  - [x] https://github.com/galaxyproject/galaxy/pull/18563
  - [x] https://github.com/galaxyproject/galaxy/pull/18586
  - [x] https://github.com/galaxyproject/galaxy/pull/18599
  - [x] https://github.com/galaxyproject/galaxy/pull/18602
  - [x] https://github.com/galaxyproject/galaxy/pull/18618
  - [x] https://github.com/galaxyproject/galaxy/pull/18645
  - [x] https://github.com/galaxyproject/galaxy/pull/18668
- [x] Fix all unit tests and mocks (seems to work in recent node versions)
- [x] Update https://github.com/galaxyproject/galaxy/blob/dev/client/docs/querying-the-api.md

---

# New way of querying the API

For more information see the [updated client docs](https://github.com/galaxyproject/galaxy/blob/9c0708d8cfb970b8313b25dd0b1f83cc445b4536/client/docs/querying-the-api.md).

```ts
import { ref, onMounted } from "vue";

import { GalaxyApi, type HDADetailed } from "@/api";
import { errorMessageAsString } from "@/utils/simple-error";

interface Props {
    datasetId: string;
}

const props = defineProps<Props>();

const datasetDetails = ref<HDADetailed>();
const errorMessage = ref<string>();

async function loadDatasetDetails() {
    // Your IDE will provide you with autocompletion for the route and all the parameters
    const { data, error } = await GalaxyApi().GET("/api/datasets/{dataset_id}", {
        params: {
            path: {
                dataset_id: props.datasetId,
            },
            query: { view: "detailed" },
        },
    });

    if (error) {
        // Handle error here. For example, you can display a message to the user.
        errorMessage.value = errorMessageAsString(error);
        // Make sure to return here, otherwise `data` will be undefined
        return;
    }

    // Use `data` here. We are casting it to HDADetailed to help the type inference because
    // we requested the "detailed" view and this endpoint returns different types depending
    // on the view. In general, the correct type will be inferred automatically using the
    // API schema so you don't need to (and shouldn't) cast it.
    datasetDetails.value = data as HDADetailed;
}

onMounted(() => {
    loadDatasetDetails();
});
```

# New way of mocking the API responses in the client unit test

Along with the openapi-fetch replacement, we introduced a new way of mocking API calls in the client using [Mock Service Worker](https://mswjs.io/).

For more information see the [updated docs](https://github.com/galaxyproject/galaxy/blob/9c0708d8cfb970b8313b25dd0b1f83cc445b4536/client/docs/unit-testing/writing-tests.md).

```ts
import { useServerMock } from "@/api/client/__mocks__";

const { server, http } = useServerMock();

describe("MyComponent", () => {
    it("should do something with the history", async () => {
        // Mock the response of the API call
        server.use(
            http.get("/api/histories/{history_id}", ({ params, query, response }) => {
                // You can use logic to return different responses based on the request
                if (query.get("view") === "detailed") {
                    return response(200).json(TEST_HISTORY_DETAILED);
                }

                // Or simulate an error
                if (params.history_id === "must-fail") {
                    return response("5XX").json(EXPECTED_500_ERROR, { status: 500 });
                }

                return response(200).json(TEST_HISTORY_SUMMARY);
            })
        );

        // Your test code here
    });
});
```


## How to test the changes?
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Upgrade openapi-typescript to 7.0.2

davelopez committed:
Update makefile to use default openapi-typescript without transform

Co-authored-by: Dannon Baker <dannon.baker@gmail.com>

davelopez committed:
Rebuild client API schema

davelopez committed:
Move openapi-typescript to dev dependencies

davelopez committed:
Replace openapi-typescript-fetch with openapi-fetch in Toolshed

davelopez committed:
Add middleware for error handling in Toolshed API client

davelopez committed:
Add openapi-fetch dependency

davelopez committed:
Refactor use openapi-fetch in some invocation related components

davelopez committed:
Replace fetcher in api/datasetCollections

davelopez committed:
Replace fetcher in api/datasets.ts

davelopez committed:
Replace fetcher in api/datatypes.ts

davelopez committed:
Replace fetcher in api/dbKeys.ts

davelopez committed:
Replace fetcher in api/forms.ts

davelopez committed:
Replace fetcher in FileSources/Instances/CreateForm.vue

davelopez committed:
Replace fetcher in ToolHelpForum.vue

davelopez committed:
Remove remaining references to ApiResponse

davelopez committed:
Update VariableValueType type to include undefined

After upgrading openapi-typescript to latest version it changed back...

davelopez committed:
Replace fetchJobDestinationParams

mvdbeek commented on client/src/api/datasetCollections.ts:
```diff
@@ -19,15 +24,19 @@ export async function fetchCollectionDetails(params: { id: string }): Promise<HD
+        },
+    });
+
+    if (error) {
+        rethrowSimple(error);
+    }
```
Could we instead register a middleware like we did with the old fetcher ?

davelopez replied:
Yes, I have considered it: https://github.com/galaxyproject/galaxy/pull/18532/files#diff-a29bb8e3afc7a891c203b6960df3ff89149f8ef3e4e80658d8a9e35a53d5620aR24

I'll work on that next. I wanted to ask about it in the meeting but couldn't make it. The thing is, I was still unsure because one of the differences between `openapi-fetch` and `openapi-typescript-fetch` is:

> openapi-typescript-fetch throws exceptions for non-OK responses (requiring you to wrap things in try/catch) rather than follow the fetch API spec of not throwing.

Something like this:
### Do not throw
```ts
const { data, error } = await client.GET("/api/libraries");
// here data can be undefined if error is defined. If you try to use it it will fail type validation
// so you are forced to handle the error at "type checking time"
if (error) {
    // handle error, return or throw
}
// data is guaranteed to be defined here
```

So if we add the middleware to rethrow like we were doing, we now probably need to wrap the call in try/catch and, at the same time, we still have to deal with `data` being possibly undefined when `error` is defined like this:

### Throw in middleware
```ts
try{
    const { data, error } = await client.GET("/api/libraries");
    // here data can still be undefined if error is defined
    if (error) {
        // handle error, return or throw
    }
    // data is guaranteed to be defined here
}catch(e){
    // handle error (again?)
}
```

Maybe there is a way to tell the middleware to assume `data` will always be defined as the exception would have been thrown otherwise?

I'm right now a bit on the fence but slightly leaning towards the first option, because it doesn't require you to remember to wrap the call in a try/catch, and on the other hand, it forces you to deal with the error early, otherwise, typescript will complain. Of course, at the cost of calling rethrowSimple or similar on the spot instead of in a single place...

mvdbeek replied:
If we're committed to handling the optional type and the error in the calling stack that would be a solid improvement for sure 👍. On a quick glance I just saw 

```
const { data, error } = await client.GET("/api/libraries");
// here data can be undefined if error is defined. If you try to use it it will fail type validation
// so you are forced to handle the error at "type checking time"
if (error) {
    rethrowSimple(error)
}
```
in the PR which is why I thought you'd opted for handling errors like we used to, in which case the middleware is less verbose.

davelopez replied:
Yes, those are some of the wrapper functions from "@api/*" that are reused in multiple places. I choose to keep those that were used in more than one place. I'm forced to rethrow there because we don't have the context to handle the error properly and the caller code should be already expecting exceptions.

There are also some situations in which error handling was not considered at all in the calling code or already had a try/catch, in those cases I'm also simply rethrowing to keep the changes minimal. In those cases where I clearly see how to handle the error I'm removing the try/catch and adapting accordingly.

davelopez committed:
Replace job lock fetchers

davelopez committed:
Replace fetcher in PageEditor/ObjectPermissions.vue

davelopez committed:
Replace fetcher in DatasetError and DatasetDetails

davelopez committed:
Replace fetcher in api/groups.ts

davelopez committed:
Replace fetcher in api/histories.archived.ts

davelopez committed:
Replace fetcher in api/histories.export.ts + refactor type imports

davelopez committed:
Replace fetcher in api/histories.ts

davelopez committed:
Replace fetcher in api/invocations.ts

davelopez committed:
Replace fetcher in api/notifications.broadcast.ts

davelopez committed:
Replace fetcher in api/notifications.preferences.ts

davelopez committed:
Replace fetcher in api/notifications.ts

davelopez committed:
Replace fetcher in api/objectStores.ts

davelopez committed:
Replace fetcher in api/pages.ts

davelopez committed:
Replace fetcher in api/quotas.ts

davelopez committed:
Replace fetcher in api/remoteFiles.ts

davelopez committed:
Replace fetcher in api/roles.ts

davelopez committed:
Replace fetcher in api/tags.ts

davelopez committed:
Replace fetcher in api/users.ts

davelopez committed:
Replace fetcher in api/workflows.ts

davelopez committed:
Refactor QuotaUsage

Expose interface and conversion function.
Simplify type handling in some components.
Consolidate duplicated API calls.

davelopez committed:
Replace fetcher in DiskUsage Management services

davelopez committed:
Replace fetcher in ObjectStore Instances services

davelopez committed:
Replace fetcher in FileSources Instances services

davelopez committed:
Replace fetcher in objectStoreTemplatesStore

davelopez committed:
Replace fetcher in objectStoreInstancesStore

davelopez committed:
Replace fetcher in fileSourceTemplatesStore

davelopez committed:
Replace fetcher in fileSourceInstancesStore

davelopez committed:
Replace fetcher in configurationStore

davelopez committed:
Replace fetcher in taskMonitor

davelopez committed:
Replace fetcher in shortTermStorageMonitor

davelopez committed:
Replace fetcher in handlesMappingJobs

davelopez committed:
Remove unused fetcher

davelopez committed:
Replace fetcher in shortTermStorage composable

davelopez committed:
Replace fetcher in visualizations grid config

davelopez committed:
Replace fetcher in visualizationsPublished grid config

davelopez committed:
Replace fetcher in pagesPublished grid config

davelopez committed:
Refactor notifications preferences components and utils

Fix type issues by splitting notification preferences into sub-components and handle updates with emits.
Other small improvements.

davelopez committed:
Replace fetcher in WorkflowList component

davelopez committed:
Fix some type inconsistencies

davelopez committed:
Consolidate imports

- Make sure `client` is always imported from "@/api"
- Use inline type import

davelopez committed:
Remove openapi-typescript-fetch dependency

davelopez committed:
Replace some axios calls with openapi-fetch client

davelopez committed:
Add `Mock Service Worker` to mock client API responses

All the jest configuration is from:
https://mswjs.io/docs/migrations/1.x-to-2.x#frequent-issues

davelopez committed:
Simplify client api exports

davelopez committed:
Refactor client API mocks and imports

davelopez committed:
Refactor Datatypes/index.test.js to use TypeScript + new client mock

davelopez committed:
Refactor client API and mock server usage

- Fix import side effects

davelopez committed:
Refactor client API calls to use GalaxyApi factory

davelopez committed:
Expose HttpResponse for mocking untyped responses in tests

davelopez committed:
Refactor HistoryArchiveExportSelector tests

To use server mock

davelopez committed:
Refactor HistoryExport component tests

To use server mock

davelopez committed:
Improve error handling in simple-error.ts

API errors are not wrapped in "data" objects when using the new fetcher.

davelopez committed:
Refactor CreateForm test to use server mock

And update object store data

davelopez committed:
Refactor HistoryImport test to use server mock

davelopez committed:
Refactor BroadcastForm test to use server mock

davelopez committed:
Refactor StsDownloadButton test to use server mock

davelopez committed:
Refactor fileSources tests to use server mock

davelopez committed:
Refactor BroadcastsList test to use server mock

davelopez committed:
Refactor NotificationsManagement test to use server mock

davelopez committed:
Refactor ExportRDMForm test to use server mock

davelopez committed:
Refactor DatasetStorage test to use TypeScript and server mock

davelopez committed:
Refactor DatasetError.test.ts to use server mock

davelopez committed:
Refactor FilesDialog.test.ts to use server mock

davelopez committed:
Refactor FormDirectory.test.js to use server mock

davelopez committed:
Refactor HistoryArchiveWizard.test.ts to use server mock

davelopez committed:
Refactor ContentItem.test.js to use server mock

davelopez committed:
Refactor SelectionOperations tests to use server mock

davelopez committed:
Fix HistoryExport/Index.test.js

It was not expecting the correct values, `expect(wrapper.exists("b-tabs-stub"))`. The exists function has no parameters and was ignoring "b-tabs-stub", it was instead asserting that the wrapper itself existed.

davelopez committed:
Add jest polyfill for clearImmediate

To avoid the following error:
ReferenceError: clearImmediate is not defined

davelopez committed:
Fix loadDbKeys function to remove unnecessary destructuring

davelopez committed:
Refactor UploadModal.test.js to use server mock

davelopez committed:
Refactor JobDestinationParams.test.ts to use server mock

davelopez committed:
Refactor Masthead.test.js to use server mock

davelopez committed:
Refactor ShowSelectedObjectStore.test.js to use server mock

davelopez committed:
Refactor UpgradeForm.test.ts to use server mock

davelopez committed:
Refactor ToolForm.test.js to use server mock

davelopez committed:
Refactor RepositoryDetails/Index.test.js to use server mock

davelopez committed:
Refactor user queries to use GalaxyApi for fetching user data

davelopez committed:
Refactor WorkflowList.test.ts to use server mock

davelopez committed:
Refactor shortTermStorageMonitor.test.ts to use server mock

davelopez committed:
Refactor taskMonitor.test.ts to use server mock

davelopez committed:
Refactor collectionAttributesStore.test.ts to use server mock

davelopez committed:
Refactor collectionElementsStore.test.ts to use server mock

davelopez committed:
Adapt adminUsers.test.js

Update "Purge" button label to "Permanently Delete"

davelopez committed:
Refactor object store templates types in the client

davelopez committed:
Remove leftover logs in tests

davelopez committed:
Fix type in broadcast notifications tests

davelopez committed:
Refactor DiskUsageSummary.test.ts to use server mock

davelopez committed:
Use correct query parameters in api/workflows

I wonder why my IDE didn't complain...

dannon committed:
Annotate globalThis in jest polyfills

dannon committed:
Add undici peer dependency

dannon committed:
Add performance, readableStream to jsdom

davelopez committed:
Fix jest fetch polyfill by using whatwg-fetch

davelopez committed:
Use fake-indexeddb in tests

Fixes:
Error in watcher callback (Promise/async): "MissingAPIError: IndexedDB API missing. Please visit https://tinyurl.com/y2uuvskb"

dannon committed:
Drop undici devDependency

dannon committed:
Drop performance polyfill, no longer necessary

davelopez committed:
Restore axios MockAdapter in ToolForm.test.js

Unfortunately there is an issue with the jest polyfills and axios requests get always an empty string "data" in response.

davelopez committed:
Convert axios queries to fetcher + type fixes

davelopez committed:
Update openapi-fetch and openapi-typescript dependencies

davelopez committed:
Update types and imports for templates

davelopez committed:
Add server mock to SelectPreferredStore test

davelopez committed:
Refactor SwitchToHistoryLink test to use server mock

davelopez committed:
Refactor RepositoryDetails tests to typescript + mock api calls

davelopez committed:
Adapt fetchItem mock response in useKeyedCache tests

davelopez committed:
Replace jest-environment-jsdom with jest-fixed-jsdom

And remove custom polyfills and whatwg-fetch.
Hopefully this would be a more stable testing environment.

davelopez committed:
Refactor SelectorModal tests to use server mock

davelopez committed:
Adapt async usage in ObjectStore CreateForm tests

davelopez committed:
Update disk usage summary test to handle pending state

davelopez committed:
Refactor UserPreferredObjectStore test to use server mock

davelopez committed:
Refactor HistoryView test to use server mock and update imports

davelopez committed:
Refactor JobInformation test to use server mock

davelopez committed:
Fix header accept value in fetchHistoryExportRecords

davelopez committed:
Update querying-the-api.md to refer to GalaxyApi client

davelopez committed:
Update writing-tests.md to include useServerMock

mvdbeek commented on client/docs/querying-the-api.md:
```diff
@@ -24,19 +24,32 @@ If there is no Composable for the API endpoint you are using, try using a (Pinia
+
+if (error) {
+    // Handle error here and return otherwise `data` will be undefined
+    return;
```
I get that's not the intention, but if someone takes this example literally that is exactly what we should avoid 😆.
Maybe we can have a real example here ?

davelopez replied:
You are right, lazy doc writing... is worse than no doc at all sometimes :sweat_smile:

davelopez replied:
Is something like https://github.com/galaxyproject/galaxy/pull/18532/commits/1ffd38242bb4926d2a0456dd05cdedd078520849 slightly better?

davelopez committed:
Update querying-the-api.md to use Toast for error handling

As a "simple" example on how to handle the API error.

davelopez committed:
Refactor DatasetStorage test to simplify API error simulation

And avoid leaving the async call unresolved

davelopez committed:
Update WorkflowSelectPreferredObjectStore.test.js

With server mock and missing API configuration endpoint. The missing API mock may cause the test to not handle the rejected promise otherwise.

davelopez committed:
Update ToolSelectPreferredObjectStore.test.js

With server mock and missing API configuration endpoint. The missing API mock may cause the test to not handle the rejected promise otherwise.

davelopez committed:
Refactor MultipleView.test.js to use server mock and update imports

mvdbeek commented on client/src/components/DatasetInformation/DatasetDetails.vue:
```diff
@@ -41,42 +41,43 @@ const datasetLoadingError = ref<string | null>(null);
+    });
+
+    if (error) {
+        jobLoadingError.value = errorMessageAsString(error);
```
This seems like a better example to me @davelopez. Toasts are simple, but they often lack context, especially if there's more than one possible error on a page.

davelopez replied:
Thanks for the suggestion, I'm adding a more complete example with some extra information about advantages/disadvantages.

davelopez replied:
Done! https://github.com/galaxyproject/galaxy/pull/18532/commits/b8be283f6ee3a6eb08186f83acd6ffc4fcae7458

mvdbeek commented on client/src/components/JobInformation/JobInformation.vue:
```diff
@@ -59,7 +59,11 @@ function filterMetadata(jobMessages) {
+            params: {
+                query: { job_id: jobId },
+            },
+        });
```
We're missing some kind of error handling here, no ?

davelopez replied:
Well spotted I'll add at least a rethrowSimple here :+1:

We can follow up with more improved error handling in general looking for all the rethrowSimple calls :sweat_smile:

davelopez replied:
Done! https://github.com/galaxyproject/galaxy/pull/18532/commits/37dcdf4a58cae25820bdb038e10dd349f8f2cf92

mvdbeek commented on client/src/components/ObjectStore/Instances/EditInstance.vue:
```diff
@@ -36,8 +37,21 @@ const loadingMessage = "Loading storage location template and instance informati
+        );
+
+        if (error) {
+            rethrowSimple(error);
```
This seems to be only handled at https://github.com/galaxyproject/galaxy/blob/dev/client/src/entry/analysis/router.js#L739-L743, if this component is ever mounted as a subcomponent there would be no error handling at all, I think ?
All of this to say maybe this needs some error handling inside of the component ?

mvdbeek replied:
Which doesn't need to happen now, it isn't worse than before.

davelopez replied:
I agree, in general, we need better error handling in most components. I didn't want to add many changes to the already monstrous PR, but we should definitely be able to search for those specific `rethrowSimple` calls and maybe transform them to better error handling in the components as a follow-up. At least it is easier to find those spots now :smiley:

mvdbeek commented on lib/tool_shed/webapp/frontend/src/components/pages/CitableRepositoryPage.vue:
```diff
@@ -18,7 +16,14 @@ const error = ref<string | null>(null)
 async function update() {
     error.value = null
-    const { data } = await indexFetcher({ owner: props.username, name: props.repositoryName })
+    const { data } = await client.GET("/api/repositories", {
```
We've lost all error handling here, no? The old middleware would have thrown on error.

davelopez replied:
Good catch again :)
Fixing it...

davelopez replied:
Done https://github.com/galaxyproject/galaxy/pull/18532/commits/094e28c3ed529820d2beae1ee62a931cccb209e6

mvdbeek commented on lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue:
```diff
@@ -26,15 +22,28 @@ async function copyKey() {
+        })
         .then(({ data }) => {
+            if (!data) {
+                throw Error("Error fetching API key")
```
Shouldn't we rethrowSimple the error here ?

davelopez replied:
I'm afraid we don't have `rethrowSimple` in the context of the Tool Shed client... I was hoping this was handled by the `notifyOnCatch` handler below. I can still refactor it for async/await syntax for better readability.

mvdbeek replied:
It (probably) is, I was more thinking that we might be missing out on valuable server-side error messages if we cast all errors to this generic string. Can't we just import rethrowSimple ?

davelopez replied:
Ahh, I see your point now! Thanks for clarifying!

davelopez replied:
AFAIK we can't import rethrowSimple because the Tool Shed 2.0 client code is now completely separated from Galaxy, but I can use the proper `error_msg` content for the error reason.

davelopez replied:
Done! https://github.com/galaxyproject/galaxy/pull/18532/commits/9c0708d8cfb970b8313b25dd0b1f83cc445b4536

mvdbeek replied:
But then we're missing out on proxy errors like gateway timeout or things like that if we assume all errors are in `err_msg`. If there's no way to share code with the Galaxy client I would copy-paste rethrowSimple (or install the galaxy client from npm ?)

davelopez replied:
Hmm... that is true, however, it seems I kept a middleware for error handling when I was experimenting with the Tool Shed client, will that cover that case?

https://github.com/galaxyproject/galaxy/pull/18532/files#diff-266bc08bc3bd8a93c2932dda21400afe346a811fb19d093b250f5925796a6416

Certainly, the Toolshed code was the first I migrated while learning, so it can benefit from some more refactoring having the new fetcher design in mind... should we give it a try in this PR or better in a follow-up?

mvdbeek replied:
The middleware approach in the TS seems fine, but we should do what Galaxy does with either rethrowSimple / errorMessageAsString. I don't think the backend is decomposed so much so that we can rely on more uniform error handling.

davelopez replied:
I've done a second round of refactoring to the Tool Shed API calls in an attempt to handle the errors consistently. I hope this is in better shape.

davelopez committed:
Add more details to querying-the-api.md

davelopez committed:
Add rethrowSimple error handling to fetchInvocation

Minimum error handling, it probably needs better error handling logic and display some feedback to the user.

davelopez committed:
Update error handling in CitableRepositoryPage.vue

davelopez committed:
Refactor Tool Shed ManageApiKey.vue to use async/await syntax

davelopez committed:
Include specific error messages in Tool Shed ManageApiKey.vue

mvdbeek commented on client/src/components/FileSources/Instances/EditInstance.vue:
```diff
@@ -36,8 +37,17 @@ const loadingMessage = "Loading file source template and instance information";
         const payload = editFormDataToPayload(template.value, formData);
-        const args = { user_file_source_id: String(instance?.value?.uuid) };
-        const { data: fileSource } = await update({ ...args, ...payload });
+        const user_file_source_id = instance?.value?.uuid!;
```
Does it compile if you do `if (template.value && instance.value)` and then remove the non-null operator ?
It seems a little clearer that way that onSubmit is only called if both of these are defined (though a comment wouldn't hurt).

davelopez replied:
Yep, that was an oversight. Thanks for the catch!

mvdbeek commented on client/src/components/History/Archiving/HistoryArchive.vue:
```diff
@@ -50,16 +48,24 @@ async function updateSearchQuery(query: string) {
+        archivedHistories.value = result.histories;
+    } catch (error) {
+        toast.error(
+            localize(`Failed to load archived histories with reason: ${errorMessageAsString(error)}`),
```
Not for this PR, but maybe we should bake that into `errorMessageAsString`

davelopez replied:
You mean the `localize` for all error messages right? :+1:

mvdbeek replied:
yes

mvdbeek replied:
I've opened https://github.com/galaxyproject/galaxy/issues/18688 with that

mvdbeek commented on client/src/components/admin/Notifications/NotificationForm.vue:
```diff
@@ -99,9 +86,37 @@ async function loadData<T>(
         target.value = tmp.map(formatter);
-    } catch (error: any) {
+    } catch (error) {
+        Toast.error(errorMessageAsString(error));
```
I think you can remove this since you're not throwing in any of the functions passed in as getData ?

davelopez committed:
Refactor EditInstance.vue to improve code readability

And remove unnecessary code

mvdbeek commented on client/src/stores/configurationStore.ts:
```diff
@@ -17,11 +15,14 @@ export const useConfigStore = defineStore("configurationStore", () => {
+                const { data, error } = await GalaxyApi().GET("/api/configuration");
+
+                if (error) {
+                    console.error("Error loading Galaxy configuration", error);
```
😢 In a follow up I would say we should add another property to the store that holds the error and makes the data possibly undefined.

davelopez committed:
Remove unnecessary try/catch in NotificationForm.vue

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

mvdbeek commented on lib/tool_shed/webapp/frontend/src/components/RegisterPage.vue:
```diff
@@ -12,19 +12,23 @@ const confirm = ref("")
+            },
         })
+        if (!data) {
+            throw new Error("No data returned")
```
That's similar to https://github.com/galaxyproject/galaxy/pull/18532#discussion_r1713624420

davelopez replied:
I think this line will never be reached because of the middleware. It could be just replaced with an empty return or a check for data is defined before using it.

mvdbeek commented on lib/tool_shed/webapp/frontend/src/components/RepositoryActions.vue:
```diff
@@ -1,10 +1,12 @@
+        .POST("/api/repositories/{encoded_repository_id}/reset_metadata", {
+            params: { path: { encoded_repository_id: props.repositoryId } },
+        })
         .catch(notifyOnCatch)
```
This only works if we keep the middleware approach (which is fine, not saying this needs to change)

mvdbeek commented on lib/tool_shed/webapp/frontend/src/components/pages/CitableRepositoryPage.vue:
```diff
@@ -18,8 +16,17 @@ const error = ref<string | null>(null)
+        },
+    })
+    if (requestError) {
+        error.value = requestError.err_msg
```
Not sure it's always a JSON response with err_msg

davelopez replied:
If we trust the schema it should :sweat_smile: 
But now that I remember I added the middleware probably none of this works as expected because all requests in the Tool Shed need to be wrapped in a try/catch, I think :thinking:

mvdbeek replied:
The schema is for the app, but the proxy can answer with gateway timeouts, rate limits etc

mvdbeek reviewed (approved):
The Galaxy portion looks great, I think for the TS side we need to not assume data.err_msg hold a proper error message.

davelopez committed:
Refactor schema imports and API calls to use ToolShedApi

This will avoid import side effects and will enable easy API mocking in unit tests in the future.

davelopez committed:
Refactor error handling in Tool Shed client

To use errorMessageAsString consistently

davelopez committed:
Refactor current user API call to use ToolShedApi

And remove unnecessary code

davelopez committed:
Remove unused axios dependency

davelopez committed:
Handle errors consistently in Tool Shed client

mvdbeek reviewed (approved):
Let's do this. Thanks a lot for working through all of the details, this is great!

dannon reviewed (approved):
This looks great to me, thanks so much @davelopez for picking this up and running with it.
